src/devices/Max7219/Max7219.cs

```diff
@@ -0,0 +1,323 @@
+        /// rotating values to the left.
+        /// </summary>
+        private void WriteBufferRotateLeft(byte[,] buffer)
+        {
+            ValidateBuffer(buffer);
+            for (int digit = 0; digit < NumDigits; digit++)
+            {
+                var mask = 0x80 >> digit;
+                var i = 0;
+                for (byte deviceId = 0; deviceId < CascadedDevices; deviceId++)
+                {
+                    _writeBuffer[i++] = (byte)((int)Register.DIGIT0 + (7 - digit));
+                    byte value = 0;
+                    byte targetBit = 0x01;
+                    for (int bitDigit = 0; bitDigit < NumDigits; bitDigit++, targetBit <<= 1)
+                    {
+                        if ((buffer[deviceId, bitDigit] & mask) != 0)
+                            value |= targetBit;
+                    }
+                    _writeBuffer[i++] = value;
+                }
+                Write(_writeBuffer);
+            }
+        }
+
+        /// <summary>
+        /// Validates the buffer dimensions.
+        /// </summary>
+        /// <param name="buffer"></param>
+        private void ValidateBuffer(byte[,] buffer)
+        {
+            if (buffer.Rank != 2)
+                throw new ArgumentException(nameof(buffer), "buffer must be two dimensional.");
+            if (buffer.GetUpperBound(0) != CascadedDevices - 1)
+                throw new ArgumentException(nameof(buffer), $"buffer upper bound ({buffer.GetUpperBound(0)}) for dimension 0 must be {CascadedDevices - 1}.");
+            if (buffer.GetUpperBound(1) != NumDigits - 1)
+                throw new ArgumentException(nameof(buffer), $"buffer upper bound ({buffer.GetUpperBound(1)}) for dimension 1 must be {NumDigits - 1}.");
+        }
+
+        /// <summary>
+        /// Clears the buffer from the given start to end (exclusive) and flushes
+        /// </summary>
+        public void Clear(int start, int end)
+        {
+            if (end < 0 || end > CascadedDevices)
+                throw new ArgumentOutOfRangeException(nameof(end));
+            if (start < 0 || start >= end)
+                throw new ArgumentOutOfRangeException(nameof(end));
+            for (int deviceId = start; deviceId < end; deviceId++)
+            {
+                for (int digit = 0; digit < NumDigits; digit++)
+                {
+                    this[deviceId, digit] = 0;
+                }
+            }
+            Flush();
+        }
+
+        /// <summary>
+        /// Clears the buffer from the given start to end and flushes
+        /// </summary>
+        public void ClearAll()
+        {
+            Clear(0, CascadedDevices);
+        }
+
+        public void Dispose()
+        {
+            if (_spiDevice != null)
+            {
+                _spiDevice.Dispose();
+                _spiDevice = null;
+            }
+        }
+
+        public enum Register : byte
```

Comment:
What we have been doing for now on other device bindings is to have the namespace be something like: Iot.Device.Max7219 and that way you can have these two enums not be nested types but instead be directly on the binding itself. That way you can also move them into different files which will make this look cleaner.